### ISSUE TITLE
Hermitpack lag issues 

### ISSUE THREAD

**Issue by NyleGarcia** [SpongePowered/SpongeForge#968]
@bloodmc @LemADEC  refrencing the hermit pack issue that wasnt resolved in ticket #948 

hermit pack warmroast  
[WarmRoasthermitpack.zip](https://github.com/SpongePowered/SpongeForge/files/535126/WarmRoasthermitpack.zip)
timings
https://timings.aikar.co/v2/?id=f11ff43a4cd447288ecfab1e069e7991

**Comment by bloodmc** [comment]
@Bebabo10 Thanks, I'll investigate this further when I get some more time to spare =)

**Comment by bloodmc** [comment]
@Bebabo10 @LemADEC @hunter688 Please test
https://www.dropbox.com/s/9ll2aa9bi58ptwu/spongeforge-1.10.2-2098-6.0.0-BETA-0.jar?dl=0

Let me know if you notice a difference. I'm troubleshooting an odd issue with this pack.

**Comment by bloodmc** [comment]
Also here is a highly experimental async build

https://www.dropbox.com/s/oy4hu3h4j3cre9w/spongeforge-1.10.2-2098-6.0.0-BETA-async-test1.jar?dl=0

Make sure you backup your entire world before testing this build. DO NOT run in production, i'm not responsible if anything breaks =)

Let me know performance results from this build.

**Comment by LemADEC** [comment]
Using
forge-1.10.2-12.18.2.2107-universal.jar
spongeforge-1.10.2-2098-6.0.0-BETA-async-test1.jar
Nucleus-0.11.0-S5-1.10.2-plugin.jar
HermitPack 1.1.0 (newest beta)

The /home and /rtp commands fail to teleport unless chunks are already loaded.
The /minecraft:tp command works.
Once I teleport to the lag base (-88 91 742), I can /home to my home base (987 78 2078), but I'm still unable to teleport to other home points.
I did /spawn, waited 2mn then tried again: home and lag base can be reached with /home but not other ones.
So I guess chunks are loading "too slow" and they are never getting unloaded.

**Comment by bloodmc** [comment]
@LemADEC Test the other build I posted as well as latest SF. The async build is highly experimental.

**Comment by LemADEC** [comment]
![image](https://cloud.githubusercontent.com/assets/4776032/19498039/5d9a7e0e-958f-11e6-8293-d6aee75add72.png)
TPS is still 20 after 90mn uptide, warmroast isn't happy.

**Comment by LemADEC** [comment]
Trying now with spongeforge-1.10.2-2098-6.0.0-BETA-1819.jar

Teleportation works as expected.
Lag is present. Looks like Forestry bee hives are reloading chunks:
![image](https://cloud.githubusercontent.com/assets/4776032/19498444/2a240da4-9591-11e6-98d9-71489f7b13c1.png)

**Comment by Vauxite** [comment]
There seems to be a chunkunloading issue.
https://timings.aikar.co/v2/?id=9e2213b1e5704b90ac0c4f7a7af1ab89

Version:
    Minecraft: 1.10.2
    SpongeAPI: 6.0.0-SNAPSHOT-73050f3
    SpongeForge: 1.10.2-2098-6.0.0-BETA-1818
    Minecraft Forge: 12.18.2.2098

**Comment by LemADEC** [comment]
After fixing Forestry beehive bugs, lag remains:
![image](https://cloud.githubusercontent.com/assets/4776032/19511287/fe29882c-95e6-11e6-84db-03f6ff672358.png)

To reproduce, I'm just doing a 10+ random teleportations to load parts of the map, including the lag base and mine.

**Comment by LemADEC** [comment]
Highlights on the Sponge part on a 10mn acquisition:
![image](https://cloud.githubusercontent.com/assets/4776032/19511577/0e168158-95e8-11e6-830a-6cbe2e46f5aa.png)
![image](https://cloud.githubusercontent.com/assets/4776032/19511591/1bb9b49c-95e8-11e6-8771-f4833e9e53f9.png)

**Comment by bloodmc** [comment]
Test the following builds and report back

https://www.dropbox.com/s/59gzwvso8gko58w/spongeforge-1.10.2-2098-6.0.0-BETA-test1.jar?dl=0
https://www.dropbox.com/s/5stru9a51zmzf5u/spongeforge-1.10.2-2098-6.0.0-BETA-test2.jar?dl=0

Let me know which build is best and if both perform better than build 1819

**Comment by NyleGarcia** [comment]
http://pastebin.com/ubAxthrM
when using test2 will share the warm roast next time we restart the build you gave us for all the mods has been running on just at certain times the server starts to drop in tps.

**Comment by NyleGarcia** [comment]
[WarmRoast.zip](https://github.com/SpongePowered/SpongeForge/files/539777/WarmRoast.zip)

**Comment by Vauxite** [comment]
@bloodmc 
Tested the 2nd one first,
https://timings.aikar.co/v2/?id=54b41389d49a4f04be4072fabe6ea2ba
Will test the other one on next restart.

**Comment by LemADEC** [comment]
With beta-test1, lag is initially severely reduced during the first 10mn profiling:
![image](https://cloud.githubusercontent.com/assets/4776032/19538989/292ab5da-9657-11e6-8541-50ca02aca253.png)
then I've restarted the profiler (still no moving from my base), and we can see it's actually lagging (11mn profiling):
![image](https://cloud.githubusercontent.com/assets/4776032/19539319/0c4c8a4a-9659-11e6-8acb-bc5e6e98c95f.png)
![image](https://cloud.githubusercontent.com/assets/4776032/19539401/87270e70-9659-11e6-8518-933f146f04e1.png)

With beta-test2, lag is initially reduced but worse than beta-test1, during first 10mn profiling:
![image](https://cloud.githubusercontent.com/assets/4776032/19539864/184cb042-965c-11e6-98db-987b0c5f2cf0.png)
after restarting the profiler, waiting 10mn, lag is gone:
https://gyazo.com/a0104645b53a64f8fa351d0c6877f7c1
![image](https://cloud.githubusercontent.com/assets/4776032/19540378/5d5ab8de-965f-11e6-950c-72036fc43639.png)

I did a few more tests and overall, beta-test2 was the only one able  to unload the laggy base (1/3 success vs 0/3), but overall beta-test1 is less laggy.

**Comment by Vauxite** [comment]
With Beta-TEST 1.
https://timings.aikar.co/v2/?id=ee04c44d92484561a29f5241068aaec4

**Comment by LemADEC** [comment]
According to our players, spongeforge-1.10.2-2098-bleeding-BETA-TEST1 is better than spongeforge-1.10.2-2098-6.0.0-BETA-1823.
In both cases, chunk-gc was disabled, if that makes any sense.

Looks like the lag increase was in fact due to a buggy tile entity from Mekanism or QuantumTanks.

**Comment by NyleGarcia** [comment]
its looking like refined storage and sponge  may be causing the issue with hermit pack https://i.gyazo.com/49f7796102cb651250fcfc0871df10b9.png

**Comment by NyleGarcia** [comment]
[WarmRoast.zip](https://github.com/SpongePowered/SpongeForge/files/545389/WarmRoast.zip)

**Comment by bloodmc** [comment]
@Bebabo10 Does that lag occur everytime you go to a specific area? If so, upload your world so I can test it. I also have your old world, if you tell me the location I can see if it occurs.

Also, please test the same area with Forge only and let me know if lag persists.

**Comment by bloodmc** [comment]
@Vauxite Looks like forestry's swarm is causing most of your lag. Is there a way you can setup a world with JUST forestry and try to reproduce the lag? I can then test it on my end.

**Comment by NyleGarcia** [comment]
#https://github.com/raoulvdberge/refinedstorage/issues/492

**Comment by NyleGarcia** [comment]
Found the issue now its at 20 tps again

**Comment by bloodmc** [comment]
@Bebabo10 Please try the same issue without Sponge and let me know if it occurs. I'm curious if Sponge is causing the lag issue.

**Comment by NyleGarcia** [comment]
https://www.dropbox.com/s/r4s2e79mjdwicak/world-hermitpack.zip?dl=1 hope you got good Internet its a 3.8gb world and will test without Forge tomorrow when player count is low

**Comment by bloodmc** [comment]
@Bebabo10 What is the lag location?

**Comment by bloodmc** [comment]
@Bebabo10 @LemADEC @Vauxite @hunter688 Please test SF 1826. It should solve the chunk unloading issue where performance slowly declines. I also optimized a few areas for chunk lookups.

**Comment by NyleGarcia** [comment]
@bloodmc What do you mean chunk lookup? and it is very difficult to find the source of the lag when your live testing ,will try to get tick profiler to work again to more easily find the locations.

**Comment by Zidane** [comment]
@Bebabo10 

Just test the newest build of SF. He wants to see if his changes make a lasting impact to your lag issues.

**Comment by NyleGarcia** [comment]
@Zidane im already testing and warmroasting with that build im responding to 

> bloodmc commented 17 hours ago • edited
> @Bebabo10 Does that lag occur everytime you go to a specific area? If so, upload your world so I can test it. I also have your old world, if you tell me the location I can see if it occurs.
> 
> Also, please test the same area with Forge only and let me know if lag persists.

**Comment by LemADEC** [comment]
Forge Mod Loader version 12.18.2.2107 for Minecraft 1.10.2
Updating from spongeforge-1.10.2-2098-bleeding-BETA-TEST1.jar
to spongeforge-1.10.2-2098-6.0.0-BETA-1826.jar
No significant lag change is observed with no chunk-gc.
Activating chunk-gc increased noticeably the lag.

However, we've major dup bug with 1826:
Some items with multiple NBT tags are remaining in inventory slots when moving around, causing instant dup, for example:
- ic2:jetpack_electric with 3 NBT tags
- weaponcase:weaponsword with 6 NBT tags

Some items are deleted instead of being dropped when clicking outside GUI (left side of the inventory):
- weaponcase:weaponsword with 6 NBT tags
- minecraft:potions (Instant Health II found in a loot chest) with 1 NBT tag

Rolling back to spongeforge-1.10.2-2098-bleeding-BETA-TEST1.jar doesn't reproduce the issue anymore.

**Comment by ghost** [comment]
Potion can confirm - its deleted when im in the survival mode - in the creative is dropped to ground. With press q works ok. Looks like something with drag&drop events...

**Comment by NyleGarcia** [comment]
> LemADEC commented an hour ago
> Forge Mod Loader version 12.18.2.2107 for Minecraft 1.10.2
> Updating from spongeforge-1.10.2-2098-bleeding-BETA-TEST1.jar
> to spongeforge-1.10.2-2098-6.0.0-BETA-1826.jar
> No significant lag change is observed with no chunk-gc.
> Activating chunk-gc increased noticeably the lag.
> 
> However, we've major dup bug with 1826:
> Some items with multiple NBT tags are remaining in inventory slots when moving around, causing instant dup, for example:
> 
> ic2:jetpack_electric with 3 NBT tags
> weaponcase:weaponsword with 6 NBT tags
> Some items are deleted instead of being dropped when clicking outside GUI (left side of the inventory):
> 
> weaponcase:weaponsword with 6 NBT tags
> minecraft:potions (Instant Health II found in a loot chest) with 1 NBT tag
> Rolling back to spongeforge-1.10.2-2098-bleeding-BETA-TEST1.jar doesn't reproduce the issue anymore.
> 
> hunter688 commented an hour ago • edited
> Potion can confirm - its deleted when im in the survival mode - in the creative is dropped to ground. With press q works ok. Looks like something with drag&drop events...

can confirm these dupes

**Comment by bloodmc** [comment]
OK thanks, ill look into the inventory issues.

**Comment by bloodmc** [comment]
Please try SF 1830 and let me know if that fixes your inventory issues.

**Comment by LemADEC** [comment]
@bloodmc We've just rebooted with spongeforge-1.10.2-2107-6.0.0-BETA-1830.jar
Still using Forge Mod Loader version 12.18.2.2107 with no chunk-gc

The dup bugs is gone but some items are unmoveable like:
- weaponcase:weaponsword with 6 NBT tags (from previous test)
- ic2:jetpack with 0 NBT tags (empty, from creative items list)
- ic2:jetpack with 1 NBT tags (fillled, from creative items list)
- ic2:jetpack_electric with 0 NBT tags (empty, from creative items list)

Noticeable are those items are moveable:
- ic2:jetpack_electric with 1 NBT tags (filled, from creative items list)
- theoneprobe:diamond_helmet_probe with 1 NBT tags (legit item or creative)
- extrautils2:BagOfHolding with 1 NBT tags (legit item, filled or empty)
- tconstruct:pickaxe with 7 NBT tags (legit item)

Throwing items from inventory no longer deletes them.

**Comment by LemADEC** [comment]
I'm still able to move/delete those items while in creative mode.

**Comment by bloodmc** [comment]
@Bebabo10 @LemADEC @Vauxite @hunter688 
Test the following build
https://www.dropbox.com/s/cy7jrrrznm0bq91/spongeforge-1.10.2-2107-6.0.0-BETA-0.jar?dl=0

Set chunk-gc-tick-interval to 1 and set chunk-unload-delay in config to 0. The main cause of lag with chunk gc enabled was the unload delay causing too many chunks to be queued for unload so setting this to 0 will force the unload to happen right away. I also made a few other improvements so let me know how it goes. As for any remaining inventory bugs, just list them.

**Comment by ghost** [comment]
potion fixed with 1830

**Comment by LemADEC** [comment]
We've updated to spongeforge-1.10.2-2107-6.0.0-BETA-0.jar
and changed chunk-unload-delay to 0

Inventory issue persist with IC2 Jetpacks and weaponcase:weaponsword. No other inventory issue in sight so far.
Lag is still present.
Profiler shows some IC2 cable reloading chunks:
![image](https://cloud.githubusercontent.com/assets/4776032/19625653/bd84df5c-991e-11e6-8494-e6d568a0110c.png)

**Comment by LemADEC** [comment]
We had a server hang with spongeforge-1.10.2-2107-6.0.0-BETA-0.jar
- There was no response on console
- Warmroast and Java debug tools were unable to connect
- kill -3 <pid> had no effect either
- forced server heap dump is available but only 28MB and fails to load in Eclipse MAT

A few minutes before that, one player was unable to move his ic2:bronze_leggings. I couldn't reproduce it with my inventory. Rebooting didn't help.

Since at least build 1830, we're seeing CME exceptions during world saving:

```
[21:43:13] [main/INFO] [mixin]: SpongePowered MIXIN Subsystem Version=0.5.17 Source=file:/home/hermitpack/mods/spongeforge-1.10.2-2107-6.0.0-BETA-1830.jar Env=SERVER
...
[01:43:27] [main/INFO] [mixin]: SpongePowered MIXIN Subsystem Version=0.5.17 Source=file:/home/hermitpack/mods/spongeforge-1.10.2-2107-6.0.0-BETA-1830.jar Env=SERVER
...
[02:06:03] [File IO Thread/ERROR]: Failed to save chunk
java.util.ConcurrentModificationException
>       at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437) ~[?:1.8.0_101]
>       at java.util.HashMap$KeyIterator.next(HashMap.java:1461) ~[?:1.8.0_101]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:23) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagList.func_74734_a(SourceFile:45) ~[dx.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagList.func_74734_a(SourceFile:45) ~[dx.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:111) ~[ea.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:101) ~[ea.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_183013_b(AnvilChunkLoader.java:242) ~[atj.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:220) [atj.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:37) [bbu.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.run(SourceFile:30) [bbu.class:?]
>       at java.lang.Thread.run(Thread.java:745) [?:1.8.0_101]
...
[05:43:42] [main/INFO] [mixin]: SpongePowered MIXIN Subsystem Version=0.5.17 Source=file:/home/hermitpack/mods/spongeforge-1.10.2-2107-6.0.0-BETA-0.jar Env=SERVER
...
[06:07:33] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 6034ms behind, skipping 120 tick(s)
[06:07:36] [File IO Thread/ERROR]: Failed to save chunk
java.util.ConcurrentModificationException
>       at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437) ~[?:1.8.0_101]
>       at java.util.HashMap$KeyIterator.next(HashMap.java:1461) ~[?:1.8.0_101]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:23) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagList.func_74734_a(SourceFile:45) ~[dx.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagList.func_74734_a(SourceFile:45) ~[dx.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:111) ~[ea.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:101) ~[ea.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_183013_b(AnvilChunkLoader.java:242) ~[atj.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:220) [atj.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:37) [bbu.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.run(SourceFile:30) [bbu.class:?]
>       at java.lang.Thread.run(Thread.java:745) [?:1.8.0_101]
[06:07:39] [Server thread/INFO]: Admin LemADEC: is anyone exploring the world right now?
...
[06:52:16] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 4658ms behind, skipping 93 tick(s)
[06:52:19] [File IO Thread/ERROR]: Failed to save chunk
java.util.ConcurrentModificationException
>       at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437) ~[?:1.8.0_101]
>       at java.util.HashMap$KeyIterator.next(HashMap.java:1461) ~[?:1.8.0_101]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:23) ~[dr.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:111) ~[ea.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:101) ~[ea.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_183013_b(AnvilChunkLoader.java:242) ~[atj.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:220) [atj.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:37) [bbu.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.run(SourceFile:30) [bbu.class:?]
>       at java.lang.Thread.run(Thread.java:745) [?:1.8.0_101]
[06:52:54] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 17279ms behind, skipping 345 tick(s)
...
[08:58:39] [main/INFO] [FML]: Loading tweaker org.spongepowered.asm.launch.MixinTweaker from spongeforge-1.10.2-2107-6.0.0-BETA-0.jar
...
[09:01:55] [Server thread/INFO]: Simpleriz lost connection: Disconnected
[09:01:55] [Server thread/INFO]: Simpleriz left the game
RemoteToolkit plugin for Bukkit was not detected. Disabling heartbeat until next restart.
[09:02:58] [File IO Thread/ERROR]: Failed to save chunk
java.util.ConcurrentModificationException
>       at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437) ~[?:1.8.0_101]
>       at java.util.HashMap$KeyIterator.next(HashMap.java:1461) ~[?:1.8.0_101]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:23) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagList.func_74734_a(SourceFile:45) ~[dx.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagList.func_74734_a(SourceFile:45) ~[dx.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:111) ~[ea.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:101) ~[ea.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_183013_b(AnvilChunkLoader.java:242) ~[atj.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:220) [atj.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:37) [bbu.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.run(SourceFile:30) [bbu.class:?]
>       at java.lang.Thread.run(Thread.java:745) [?:1.8.0_101]
[09:03:01] [Server thread/INFO] [nucleus]: 2muChPride ran the command: /home ha
[09:03:03] [Server thread/ERROR] [Sponge]: No entity type is registered for class crazypants.enderio.machine.killera.TileKillerJoe$Attackera
...
[09:03:18] [Server thread/INFO]: BLK_WARRIOR joined the game
[09:03:18] [Server thread/INFO] [STDOUT]: [mcjty.lib.preferences.PreferencesProperties:syncToClient:31]: syncToClient: style = STYLE_FLAT_GRADIENT
[09:03:19] [File IO Thread/ERROR]: Failed to save chunk
java.util.ConcurrentModificationException
>       at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437) ~[?:1.8.0_101]
>       at java.util.HashMap$KeyIterator.next(HashMap.java:1461) ~[?:1.8.0_101]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:23) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagList.func_74734_a(SourceFile:45) ~[dx.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagList.func_74734_a(SourceFile:45) ~[dx.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_150298_a(NBTTagCompound.java:453) ~[dr.class:?]
>       at net.minecraft.nbt.NBTTagCompound.func_74734_a(NBTTagCompound.java:26) ~[dr.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_150663_a(CompressedStreamTools.java:111) ~[ea.class:?]
>       at net.minecraft.nbt.CompressedStreamTools.func_74800_a(CompressedStreamTools.java:101) ~[ea.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_183013_b(AnvilChunkLoader.java:242) ~[atj.class:?]
>       at net.minecraft.world.chunk.storage.AnvilChunkLoader.func_75814_c(AnvilChunkLoader.java:220) [atj.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.func_75736_b(SourceFile:37) [bbu.class:?]
>       at net.minecraft.world.storage.ThreadedFileIOBase.run(SourceFile:30) [bbu.class:?]
>       at java.lang.Thread.run(Thread.java:745) [?:1.8.0_101]
[09:03:22] [Server thread/WARN] [Sponge]: Unable to find InventoryPacketState handler for click window packet: CPacketClickWindow{windowId=5, slotId=-999, usedButton=4, actionNumber=1, clickedItem=null, mode=QUICK_CRAFT}
[09:03:24] [Server thread/WARN]: Can't keep up! Did the system time change, or is the server overloaded? Running 5498ms behind, skipping 109 tick(s)
...
```

Those exceptions won't show up in log files saved by Forge, for unknown reason, so I'm unable to confirm when it started.

We've rolled back to the API 5 bleeding version for now.

**Comment by gabizou** [comment]
Just going to mention that the API 5 bleeding version is not the same as API 5 stable version. 

Another note: I do have a distaste for Torcherino, mostly because it masks performance by calling additional tile updates:

![torcherino](https://puu.sh/rTxuK/c670410bd5.png)

As an example, this is calling neighboring blocks updates, including blocks that generally are unperformant, even if harmless (such as grass or snow).

**Comment by bloodmc** [comment]
@Bebabo10 @LemADEC @Vauxite @hunter688 Please test SF 1835 and report back. Also make sure to test with chunk-gc-tick-interval enabled as I've fixed the major issue causing slowdowns with it. Feel free to experiment as well with chunk-unload-delay between 0-30.

**Comment by NyleGarcia** [comment]
@bloodmc fixed in build 1835

**Comment by LemADEC** [comment]
For the record, Torcherino was disabled in all my testings.

Lag base still remains chunkloaded, lag is getting better but it's still present on this 2 mn profiling:
![image](https://cloud.githubusercontent.com/assets/4776032/19636645/d1ecbc7e-99ca-11e6-97c6-5c1bf139fdf3.png)
![image](https://cloud.githubusercontent.com/assets/4776032/19636653/e830f9c8-99ca-11e6-992d-ef1462bd25f3.png)
This is with chunk-gc interval 1 and delay 0.

Inventory issue with Jetpack is still present => is there a separate issue to track it?

**Comment by NyleGarcia** [comment]
nvm #981  https://timings.aikar.co/v2/?id=98743f8f4ccb43c0b8f7ef3109cc9c83&start=1477331947&end=1477332061&section=lag IC2 tile entities and refined storage again

**Comment by bloodmc** [comment]
@LemADEC TrackingUtil simply wraps around each TE tick so we can capture relevant information and the real slowdown is with whatever TE it is showing so I would report to the mod dev as that is not a sponge issue.

Report any performance issues with refined storage here

https://github.com/raoulvdberge/refinedstorage/issues

**Comment by LemADEC** [comment]
@bloodmc My concern is about java.utils.ArrayList.removeAll taking 14% CPU. To me this indicates chunks are in constant reloading.
On server we had to disable chunk-gc again because lag was way worse with it (more than 150ms per tick).
I wish we had a mod/plugin monitoring 'fast chunk reloads' to help identify such bogus behavior.

Have we considered reimplementing the Cauldron (?) trick that disables TE update ticks when no players are in vicinity?

**Comment by bloodmc** [comment]
@LemADEC Set the following in your config

mob-spawn-range 4
chunk-gc-tick-interval 1
chunk-unload-delay 0
max-chunk-unloads-per-tick=100
auto-save-interval 6000

If the issue still occurs, turn on debug logging by setting the following to true
chunk-load=true
chunk-unload=true
log-stacktraces=true

Once this is done, simply execute
/sponge reload -g
This will refresh the global config.
Note: This will cause major spam as chunks load/unload but it will show exactly whats causing it. When done with debug, simply set them back to false and run the same command again.

Also upload your world and let me know the location where this lag is occuring and provide a timings report (not warmroast).

**Comment by LemADEC** [comment]
@bloodmc  timings isn't a valid sub command when I type "/sponge timings" or "/sponge timings -g"

To reproduce quickly, I go the lag base (same world and coordinates I've posted last week: -99 91 753), then do 20 times /rtp (using Nucleus plugin, with 30 attempts at 3000 radius), then go back to my base.

**Comment by bloodmc** [comment]
@LemADEC For sponge timings the command would be

/sponge timings on
wait a few minutes
/sponge timings paste

When done with timings,
/sponge timings off

I still need the output from the chunk load/unload debug. Follow what i said above and report back with the logs.

**Comment by LemADEC** [comment]
@bloodmc Server console logger crashed due to the spam so I couldn't get the timings. Here's the logs: 
It probably crashed too soon after enabling logs to see much, I'll try again later this week with a larger buffer.

**Comment by LemADEC** [comment]
Please remove the link once you've got it.

**Comment by bloodmc** [comment]
@LemADEC Thanks, I was able to reproduce the issue. Working on a fix.

**Comment by bloodmc** [comment]
@LemADEC Turns out some logic was wrong in my chunkgc handler. I'll be pushing a fix shortly =) Looks like everything is working perfectly now, just need to test a bit more.

**Comment by bloodmc** [comment]
@LemADEC Please test latest build and confirm that the lag is fixed. I would recommend keeping the chunk delay at around 15. There were problems with this in earlier builds due to the ChunkGC logic not being correct. I also introduced some optimizations to fight against mods that leak chunks.

**Comment by bloodmc** [comment]
I'm pushing a better build, so just wait for next one.

**Comment by bloodmc** [comment]
@LemADEC @Bebabo10 @Vauxite @hunter688 Please test SF 1841. It should solve this issue once and for all =). Before you run the build, make sure you change the following defaults in config

chunk-gc-tick-interval - 600
chunk-unload-delay - 15

The chunk gc will now run once every 30 seconds to cleanup any chunks that leaked. The chunk unload delay is only for player chunks (doesn't affect leaks). So if a chunk that was used by a player is not used within 15 seconds, it unloads. If it gets active before unload, the timer resets.

**Comment by LemADEC** [comment]
The lag appears to be gone for the most part. Here's a timings report about
10mn after my lag stress test:
https://timings.aikar.co/v2/?id=8dff316c281143f988b370fb5217590e

Sadly, we're observing a lot of IC2 crashes since we've deployed the latest
fix. The pattern was there before when players where dismounting IC2
consumers. It's a bit too soon to conclude if the last sponge changes are
causing it with the chunk gc. We've disabled the latest for now and will
see how it goes. Here's the typical pattern observed:
'''
java.lang.NullPointerException: Exception in server tick loop
        at
ic2.core.energy.leg.EnergyNetLocalLeg.emitEnergyFrom(EnergyNetLocalLeg.java:341)
        at
ic2.core.energy.leg.EnergyNetLocalLeg.tickEnd(EnergyNetLocalLeg.java:144)
        at ic2.core.energy.leg.EventHandler.preTick(EventHandler.java:46)
        at
net.minecraftforge.fml.common.eventhandler.ASMEventHandler_189_EventHandler_preTick_WorldTickEvent.invoke(.dynamic)
        at
net.minecraftforge.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:90)
        at
net.minecraftforge.fml.common.eventhandler.EventBus.post(EventBus.java:130)
        at
net.minecraftforge.fml.common.eventhandler.EventBus.post(EventBus.java:99)
        at
net.minecraftforge.fml.common.FMLCommonHandler.onPostWorldTick(FMLCommonHandler.java:269)
        at
net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:718)
        at
net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:387)
        at
net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:613)
        at
net.minecraft.server.MinecraftServer.run(MinecraftServer.java:471)
        at java.lang.Thread.run(Thread.java:745)
'''

2016-10-25 19:43 GMT+02:00 bloodmc notifications@github.com:

**Comment by bloodmc** [comment]
@Player any idea why the above would happen so I can check our end?

**Comment by bloodmc** [comment]
@sfPlayer any idea why the above would happen so I can check our end?

**Comment by Aroma1997** [comment]
That crash is caused by an IC2 addon not using the API correctly. My guess: calculator.
In a newer build of IC2, we changed that to log the error instead of crashing the game. Try updating your IC2 version.

**Comment by LemADEC** [comment]
This is from the FTB HermitPack 1.1.0 version, Calculator isn't part of that pack. However TechReborn and Mekanism are in.
Will the log give a block name or at least a mod name so we know where to look for?

**Comment by Aroma1997** [comment]
If you update your IC2 version, it will.
